archive/snappy.ksy

```diff
@@ -0,0 +1,57 @@
+meta:
+  id: snappy
+  title: snappy framing format
+  license: CC0-1.0
+  ks-version: 0.9
+  encoding: utf-8
+  endian: le
+doc-ref: https://github.com/google/snappy/blob/master/framing_format.txt
+doc: |
+  snappy's framing2 format
+  
+  Test files can be created with snzip: <https://github.com/kubo/snzip>
+seq:
+  - id: chunks
+    type: chunk
+    repeat: until
+    repeat-until: _io.eof or not _.is_valid
+    # This is ugly, as it eats one extra byte, so an external
+    # program processing this could should take this into account
+types:
+  chunk:
+    seq:
+      - id: identifier
+        type: u1
+        enum: chunk_types
+      - id: body
+        type: chunk_body
+        if: is_valid
+    instances:
+        # more values are actually allowed but in practice
+        # these aren't encountered
+        is_valid:
+          value: identifier == chunk_types::compressed or
+            identifier == chunk_types::uncompressed or
+            identifier == chunk_types::padding or
+            identifier == chunk_types::stream_identifier
+  chunk_body:
+    seq:
+      - id: len_chunk
+        type: int3
+      - id: data
+        size: len_chunk.value
+  int3:
+    seq:
+      - id: lower
+        type: u2
+      - id: higher
+        type: u1
+    instances:
+      value:
+        value: higher * 65536 + lower
```

Comment:
I usually use bit ops here.

Reply:
> I usually use bit ops here.

Yes, that would definitely make more sense here. It's about expressing the intent: here `higher * 65536` is actually meant to be `higher << 16` (or even `higher << lower._bitsizeof` which is supported [since 0.9](https://kaitai.io/news/2020/10/16/kaitai-struct-v0.9-released.html#release-highlights)) and `+ lower` should be `| lower`. This is quite literally what you want to do - first shift bits of `higher` (`00 00 HH`)  to the left (`HH 00 00`) so that it doesn't collide with `lower` (`00 LL LL`) and then compose them together to get `HH LL LL` - the bitwise OR is enough for that, there is no logical reason why you would want addition `+` with all that carry-in/carry-out madness.

_(This review comment is probably made obsolete by https://github.com/kaitai-io/kaitai_struct_formats/pull/544/files#r744576387 in the sense that the type `int3` would be dropped, but I can't miss the opportunity to share my view on this :-) )_

Reply:
> (or even `higher << lower._bitsizeof` which would work [since 0.9](https://kaitai.io/news/2020/10/16/kaitai-struct-v0.9-released.html#release-highlights))

OK, it is broken:

```ksy
meta:
  id: test
  endian: le
seq:
  - id: i
    type: int3
types:
  int3:
    seq:
      - id: lower
        type: u2
      - id: higher
        type: u1
    instances:
      value:
        # value: higher * 65536 + lower
        value: higher << lower._bitsizeof | lower
```

```
/types/int3/instances/value/value: error: don't know how to call method '_bitsizeof' of object type 'IntMultiType(false,Width2,Some(LittleEndian))'
```

and that is just another reminder of how much work there is to be done.

But at least `_sizeof` works:

```ksy
    instances:
      value:
        # value: higher * 65536 + lower
        value: higher << (8 * lower._sizeof) | lower
```

though it's slightly less elegant.